include/tvm/relay/ad.h

```diff
@@ -0,0 +1,48 @@
+/*!
+ *  Copyright (c) 2018 by Contributors
+ * \file ad.h
+ * \brief Common api shared by all the automatic differentiation(AD) algorithm.
+ */
+#ifndef TVM_RELAY_AD_H_
+#define TVM_RELAY_AD_H_
+
+#include <tvm/relay/expr.h>
+#include <tvm/relay/module.h>
+
+namespace tvm {
+namespace relay {
+
+/*! What is automatic differentiation(AD) and why is it important?
+ * By AD, we roughly mean, given a term which denote some mathematical function,
+ * derive a term which denote the derivative of that mathematical function.
+ * Such a method can be compile-time, which is a macro on completely known function.
+ * Formally speaking, such requirement mean that the input function is a closed expression -
+ * that is, it only refer to local variable that is it's parameter, or defined inside it.
+ * Every top level definition satisfy this criteria.
+ * AD can also be run-time, which mean it is merely a function term of AD : (Float[] -> Float[]) -> (Float[] -> Float[]).
+ * In relay we currently only support compile-time AD, but it should be enough for a lot of use case.
+ *
+ * In deep learning, the most common way to train a deep neural network is by gradient descend or some of it's variant.
+ * Such optimization method require us to input the gradient of neural network, which can be obtained easily using AD.
+ * In fact, back propagation is essentially reverse-mode automatic differentiation, a kind of AD!
+ */
+
+/*! In relay, automatic differentiation(AD) is a macro,
+ *  that transform closed expr(expr without free variable/free type variable) of type
+ *  (x0, x1, x2, ...) -> Float[] to
+ *  (x0, x1, x2, ...) -> (Float[], (x0, x1,  x2, ...)),
```

Comment:
Could you please clarify the meaning of first tuple item `Float[]` in gradient type `(x0, x1, x2, ...) -> (Float[], (x0, x1,  x2, ...))`? Is it a result value of the original function?